### PR TITLE
many: add authcontext to handle user updates to state

### PR DESCRIPTION
**Pull request by matiasb**
Add AuthContext providing UpdateUser method to update a user state. AuthContext is passed to StoreService on creation (and will help updating discharge macaroons on refresh).

### CODE REVIEW

File: overlord/auth/auth.go [hunk 1]
```diff
@@ -183,3 +183,41 @@ NextUser:
...
 	return nil, ErrInvalidAuth
 }
+
+// An AuthContext handles user updates.
+type AuthContext interface {
+	UpdateUser(user *UserState) error
+}
+
+// AuthStateContext helps keeping track and updating users in the state.
+type AuthStateContext struct {
```
Comment (niemeyer) [inline overlord/auth/auth.go]: Let's please define this as `type authContext struct` and make it private to this package. Externally it'll be known via its interface only.

File: overlord/auth/auth.go [hunk 2]
```diff
@@ -183,3 +183,41 @@ NextUser:
...
+	state *state.State
+}
+
+// NewAuthContext returns an AuthContext for state.
+func NewAuthStateContext(st *state.State) *AuthStateContext {
```
Comment (niemeyer) [inline overlord/auth/auth.go]: `NewAuthContext`.. doc is already right.

Comment (niemeyer) [inline overlord/auth/auth.go]: (and return the interface, not the concrete type)

File: overlord/auth/auth.go [hunk 3]
```diff
@@ -183,3 +183,41 @@ NextUser:
...
+}
+
+// UpdateUser updates user in state.
+func (asc *AuthStateContext) UpdateUser(user *UserState) error {
+	var authStateData AuthState
+
+	asc.state.Lock()
+	defer asc.state.Unlock()
+
+	err := asc.state.Get("auth", &authStateData)
```
Comment (niemeyer) [inline overlord/auth/auth.go]: Let's please move this logic to the `auth.UpdateUser(st, user)` function, and call it from here. It's a good partner to existing logic (NewUser, User, UpdateUser, RemoveUser).

**Comment by niemeyer** [comment]
A few comments and LGTM.

**Comment by chipaca** [comment]
:+1: 

adding an argument to a constructor that already takes a config is weird, but we're abusing the nil config in that one so fixing it would require a refactor beyond the scope of this.

**Comment by pedronis** [comment]
@chipaca actually AuthContext carries a state so using the config to carry it doesn't seem that good either anyway

**Pull request merged**